man/avg_videoinfo.1
.TH AVG_VIDEOINFO "1" "March 2011" "libavg 1.5.4" "User Commands"
.SH NAME
avg_videoinfo \- video inspection tool
.SH SYNOPSIS
.B avg_videoinfo
\fIfile(s)\fR \fI[OPTION]\fR
.SH DESCRIPTION
Dump video file format information to the console
.PP
Inspects the given file(s) and dumps information about duration, bitrate, video/audio codec,
video frame size and pixel format, video fps, audio sample rate and number of channels for
each file given.
.PP
The default output format is text.
.PP
XML and CSV formats can be chosen with the appropriate options.
.PP
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a brief description of the options.
.TP
\fB\-x\fR, \fB\-\-xml\fR
XML output format.
.TP
\fB\-c\fR, \fB\-\-csv\fR
CSV output format.
.SH AUTHOR
Ulrich von Zadow <coder@c-base.org>


NEWS
libavg 1.5.4 - February 28, 2011 (Linux only)

  svn revision 5644, branches/release1.5.x

  XInput 2.1 updates/fixes.

libavg 1.5.3 - February 20, 2011 (Linux only)

  svn revision 5599

  Major new feature:
    * Linux XInput 2.1 Multitouch support.

libavg 1.5.1 - November 14, 2010 (Mac only)

  svn revision 5275

  Fixed bug that caused fullscreen apps to crash on the mac (Bug # 162).

libavg 1.5.0 - November 6, 2010

  svn revision 5270

  Major New Features:
    * Nodes can be constructed directly from python, making createNode unnecessary.
    * Added support for rendering avg scenes to offscreen canvases.
    * Added GPU-based effects framework and Blur, Shadow, Chromakey effects.
    * Added multitouch driver model and support for Apple Magic Trackpads and TUIO 
      devices.
    * Support for flash videos with an alpha channel. Decoding is GPU-accellerated.

  Other Changes:
    * Added color controls (gamma, brightness, contrast) to all RasterNodes.
    * Node constructors have a parent parameter which makes a following addChild 
      redundant.
    * Added Player.setEventHook().
    * Added onscreen keyboard for touchscreens.
    * WordsNode.getNumLines(), .getCharIndexFromPos(), .getTextAsDisplayed().
    * DivNode.getEffectiveMediaDir().
    * Clearer profiling that works for multiple threads.
    * Added AVG_BREAK_ON_IMPORT and AVG_BREAK_ON_ASSERT env variables.
    * Added AVG_DUMP_TEST_FRAMES env variable.
    * Added Player.setWindowFrame().
    * Added VideoNode.queuelength.
    * Video stability & timing improvements.
    * DivNode crop default is now false.
    * Node class names are unified: All names end with 'Node': WordsNode, DivNode, etc.
    * Added ImageNode.compression for 16 bpp texture support.
    * Added avg_audioplayer.py.

    * Various compiler compatibility fixes: Compiles with gcc 4.5 and VC++ 2010 now.
    * Lots of bugfixes - see the bug tracker.

libavg 1.5.0.pre3 - November 4, 2010

  svn revision 5267

libavg 1.5.0.pre2 - November 3, 2010

  svn revision 5263

libavg 1.5.0.pre1 - October 31, 2010

  svn revision 5244 

libavg 1.0.0 - January 8, 2010

  svn revision 4489

  New Features:

    * hinting is now optional in words node.
    * Multitouch emulation module added to AVGMTApp.
    * masks now work with words nodes.
    * Added video.volume property and video.hasAudio() method.
    * Added AVG_LOG_CATEGORIES environment variable. Log configuration is now
      done using the environment and not by calling a method.
    * Added maskpos and masksize attribute so masks can be moved and rescaled.
    * Added Player.getKeyModifierState()
    * Added sse2 version of yuv->rgb color conversion.
    * Added avg_videoinfo.sh.
    * Added mesh node.
    * Added avg_showcamera --resetbus to reset the firewire bus after crashes.

  Other Improvements:
    * New anim framework is now stable.
    * Snow Leopard compatibility.
    * avg_showcamera has much better command line handling.
    * The test suite is faster and a lot more comprehensive.
    * hrefs and mediadirs are now unicode-aware, so filenames like "ö.png" 
      should work.
    * node.unlink(True) now deletes all event handlers in the node, making
      it a lot easier to avoid memory leaks. 
    * Lots of bug fixes:
      - Windows gamma is now reset correctly on program end.
      - Linux vsync now much more stable.
      - Setting node.pos.x now throws an error instead of being ignored. 
        Same for all other Point2D attributes.
      - video.getNumFrames() now returns the exact number of frames.
      - Behaviour of nodes before Player.play() and when they are not in the
        avg tree is now much more stable.
      - Fixed words node sensitive area for centered and right-aligned nodes.

libavg 1.0.0.pre3 - January 6, 2010

  svn revision 4475

libavg 1.0.0.pre2 - December 23, 2009

  svn revision 4398

libavg 1.0.0.pre1 - December 21, 2009

  svn revision 4387

libavg 0.9.0 - August 23, 2009

  svn revision 4100

  Major new features:

    * Support for vector node types: Line, PolyLine, Curve, Circle, Polygon,
      Rectangle. Vector nodes can be textured, with the closed vectors taking two
      textures: one for the outline, one for the fill. Vector drawing is GPU-based.
    * Plugin support. Plugins written in C++ can define complete new node types.
    * Powerful (but still experimental) new animation framework.
    * Unified & much more stable camera drivers. Consistent support for all 
      resolutions & framerates we could test. Support for more than one camera on all
      platforms. fw800 is supported.
    * Uniform support for aggregate types in node interfaces (node.pos, node.pivot, 
      ...)

  Other improvements:
    * gcc 4.4 and ffmpeg 0.5.0 compatibility.
    * Added letterspacing support to words node.
    * Threaded videos are now the default.
    * Improved video seek and loop performance.
    * New Node::getElementByPos() method that returns the node at that point on the
      screen.
    * Added avg.svnrevision and avg.svndate properties so the avg version can be 
      queried.
    * New avg_showfont.py utility to help font selection.
    * New avg_showcamera.py utility to help camera setup.
    * New avg.getMemoryUsage function that returns the resident set size.
    * New Words.addFontDir().
    * Added support for mipmapping node textures, including videos.
    * rasternode.maskhref to support static alpha masks for any videos, word nodes,
      etc.

  Syntax and semantic changes:
    * In the course of adding aggregate types for node attributes, several attributes
      have changed:
      - words.size is now called words.fontsize.
      - pivotx and pivoty have been replaced by pivot.
    * right-aligned and centered words nodes are positioned differently.
    * avgtrackerrc syntax has changed significantly; see src/avgtrackerrc.minimal.
    * Support for firewire cameras using libdc1394 ver. 1 has been phased out. Please 

src/utils/avg_videoinfo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# libavg - Media Playback Engine.
# Copyright (C) 2003-2008 Ulrich von Zadow
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Current versions can be found at www.libavg.de
#

from optparse import OptionParser, OptionValueError
import sys, traceback
from libavg import avg
import time
from xml.dom import minidom
from xml import dom
import os

parser = OptionParser("usage: %prog <videofilename> [options]")
parser.add_option("-x", "--xml", dest = "xml", action="store_true",
        help = "Set to enable output as xml")
parser.add_option("-c", "--csv", dest = "csv", action="store_true",
        help = "Set to enable output as csv")
options, args = parser.parse_args()

Player = avg.Player.get()

Player.loadString("""
    <?xml version="1.0"?>
    <!DOCTYPE avg SYSTEM "../../doc/avg.dtd">
    <avg width="1280" height="720">
        <video id="video" x="0" y="0" threaded="true"/>
    </avg>
""")
node = Player.getElementByID("video")
impl = minidom.getDOMImplementation()
doc = impl.createDocument(None, "videodict", None)
rootElement = doc.documentElement
CSV_video = ''
VideoList = []
len_filename = 0

def appendXMLChild(node):    
    node.play()
    videoinfo = doc.createElement("videoinfo")
    videoinfo.setAttribute("file", node.href)
    videoinfo.setAttribute("duration", str(node.getDuration()/1000.))
    videoinfo.setAttribute("bitrate", str(node.getBitrate()))
    rootElement.appendChild(videoinfo)
    videoNode = doc.createElement("video")
    videoNode.setAttribute("codec", node.getVideoCodec())
    videoNode.setAttribute("size", str(node.getMediaSize()))
    videoNode.setAttribute("pixelformat", node.getStreamPixelFormat())
    videoNode.setAttribute("fps", str(node.fps))
    videoinfo.appendChild(videoNode)
    if node.hasAudio():
        audioNode = doc.createElement("audio")
        audioNode.setAttribute("codec", node.getAudioCodec())
        audioNode.setAttribute("samplerate", str(node.getAudioSampleRate()))
        audioNode.setAttribute("channels", str(node.getNumAudioChannels()))
        videoinfo.appendChild(audioNode)

def singleVideoInfo(node):
    node.play()
    print "File: " + node.href
    print ("Duration: " + str(node.getDuration()/1000.) + " s (" 
            + str(node.getNumFrames()) + " frames)")
    print "Bitrate: " + str(node.getBitrate()) + " b/s"
    print "Video stream: " 
    print "  Codec: " + node.getVideoCodec()
    print "  Size: " + str(node.getMediaSize()) + " pixels"
    print "  Pixel format: " + node.getStreamPixelFormat()
    print "  FPS: " + str(node.fps)
    if node.hasAudio():
        print "Audio stream: " 
        print "  Codec: " + node.getAudioCodec()
        print "  Sample rate: " + str(node.getAudioSampleRate()) + " Hz"
        print "  Number of channels: " + str(node.getNumAudioChannels())
  
def CSVtable(node):
    global CSV_video
    
    node.play()
    if CSV_video == '':
        CSV_head = ("File\tDuration\tBitrate\tVideoCodec\t" + \
                "VideoSize\tPixel format\tFPS\tAudioCodec\t" + \
                "Audio Sample rate\t Audio channels"+"\n")
        CSV_video = CSV_head   
    CSV_video += \
        str(node.href)+'\t'+\
        str(node.getDuration()/1000.) + " s (" + str(node.getNumFrames()) + \
            " frames)"+'\t'+\
        str(node.getBitrate()) + " b/s" + '\t' + \
        str(node.getVideoCodec()) + '\t' + \
        str(node.getMediaSize()) + " pixels" + '\t' + \
        str(node.getStreamPixelFormat()) + '\t' + \
        str(node.fps) + '\t'
    if node.hasAudio():
        CSV_video += \
            str(node.getAudioCodec()) + '\t' + \
            str(node.getAudioSampleRate()) + " Hz" + '\t' + \
            str(node.getNumAudioChannels()) + '\n'
    else:
        CSV_video += ' \t \t \n'
    
def showInfo(node):    
    global len_filename
    node.play()
    if node.hasAudio():
        vFile = str(os.path.basename(node.href)).ljust(len_filename+2)
        vDuration = str(node.getDuration()/1000.).ljust(12)
        vBitrate = str(node.getBitrate()).ljust(15) 
        vVideoCodec = str(node.getVideoCodec()).ljust(13)
        vVideoSize = str(node.getMediaSize()).ljust(11)
        vPixel = str(node.getStreamPixelFormat()).ljust(14)
        vFPS = str(round(node.fps, 2)).ljust(10)
        vAudioCodec = str(node.getAudioCodec()).ljust(12)
        vSampleRate = str(node.getAudioSampleRate()).ljust(13)
        vChannels = str(node.getNumAudioChannels()).ljust(8)
        
        info = vFile + vDuration + vVideoCodec + vVideoSize + vPixel + vFPS + vAudioCodec + vSampleRate + vChannels
        
        print info
    else:
        vFile = str(os.path.basename(node.href)).ljust(len_filename+2)
        vDuration = str(node.getDuration()/1000.).ljust(12)
        vBitrate = str(node.getBitrate()).ljust(15) 
        vVideoCodec = str(node.getVideoCodec()).ljust(13)
        vVideoSize = str(node.getMediaSize()).ljust(11)
        vPixel = str(node.getStreamPixelFormat()).ljust(14)
        vFPS = str(round(node.fps,2)).ljust(10)
        vAudioCodec = "".ljust(12)
        vSampleRate = "".ljust(13)
        vChannels = "".ljust(8)
        
        info = vFile + vDuration + vVideoCodec + vVideoSize + vPixel + vFPS + vAudioCodec + vSampleRate + vChannels
        
        print info
        
    
            
if len(sys.argv) ==1:
    parser.print_help()
    sys.exit(1)

if len(args) == 1:
    node.href = sys.argv[1]
    if options.xml:
        appendXMLChild(node)
        print doc.toprettyxml(indent="    ",encoding="utf-8")
    elif options.csv:
        CSVtable(node)
        print CSV_video
    else:
        singleVideoInfo(node)
        
else:
    for file in args: 
        try:
            node.href = str(os.path.abspath(str(file)))    
            #node.href = str(os.path.join(os.path.normpath(os.getcwd() + "/" +str(file))))
            node.play()
            VideoList.append(node.href)   
        except:
            sys.stderr.write("Error in getting Videoinfo: " + str(node.href) + "\n")
    
    for i in xrange(0, len(VideoList)):
        if len_filename < len(str(os.path.basename(str(VideoList[i])))):
            len_filename = len(str(os.path.basename(str(VideoList[i]))))
            
    for i in xrange(0, len(VideoList)):
        node.href = str(VideoList[i])
        if options.xml:
            appendXMLChild(node)
        elif options.csv:
            CSVtable(node)
        else:
            if i == 0:
                vFile = "File".ljust(len_filename+2)
                vDuration = "Duration[s]".ljust(12)
                vBitrate = "Bitrate [b/s]".ljust(15) 
                vVideoCodec = "VideoCodec".ljust(13)
                vVideoSize = "VideoSize".ljust(11)
                vPixel = "Pixel format".ljust(14)
                vFPS = "FPS".ljust(10)
                vAudioCodec = "AudioCodec".ljust(12)
                vSampleRate = "Sample rate".ljust(13)
                vChannels = "Channels".ljust(8)
                
                title = vFile + vDuration + vVideoCodec + vVideoSize + vPixel + vFPS + vAudioCodec + vSampleRate + vChannels
                
                print title
                
                showInfo(node)
            else:    
                showInfo(node)
    if options.xml:
        print doc.toprettyxml(indent="    ",encoding="utf-8")
        
    elif options.csv:
        print CSV_video
    
    

        
